certbot-nginx/certbot_nginx/configurator.py

```diff
@@ -250,6 +259,121 @@ def _get_ranked_matches(self, target_name):
                                 'rank': 6 if vhost.ssl else 7})
         return sorted(matches, key=lambda x: x['rank'])
 
+    def all_vhost_ports(self, domain):
+        """Get all ports that all vhosts are listening on, that match domain.
+
+        :param str domain: domain to find ports of
+        :returns: set of strings representing the port numbers
+        :rtype: set
+        """
+        matching_vhosts = [item['vhost'] for item in self._get_ranked_matches(domain)]
+        out = set([addr.get_port() for vhost in matching_vhosts for addr in vhost.addrs])
+        # also check the default port, 80
+        out.add(self.DEFAULT_PORT)
+        if "" in out:
+            out.remove("")
+        if r"*" in out:
+            out.remove(r"*")
+        return out
+
+    def choose_redirect_vhost(self, target_name, port):
+        r"""Chooses a virtual host for redirect enhancement.
```

Comment:
Don't worry too much about getting the docs to display perfectly. It was just something small I noticed, however, here's how to see the problem.

I saw it using `IPython`. If you've [activated the virtual environment](https://certbot.eff.org/docs/contributing.html#getting-started) and have `IPython` installed, you can do:
```
$ ipython
Python 2.7.12 (default, Jul  1 2016, 15:12:24) 
Type "copyright", "credits" or "license" for more information.

IPython 5.1.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: from certbot_nginx.configurator import NginxConfigurator

In [2]: NginxConfigurator.choose_redirect_vhost?
Signature: NginxConfigurator.choose_redirect_vhost(self, target_name, port)
Docstring:
Chooses a virtual host for redirect enhancement.

Filter all hosts for those that include listen port but do not include
ssl on. Of these, find the most matching one based on target_name.

.. todo:: This should maybe return list if no obvious answer
    is presented.

.. todo:: The special name "$hostname" corresponds to the machine's
    hostname. Currently we just ignore this.

:param str target_name: domain name
:param str port: port number or "\*" or ""
:returns: vhost associated with name
:rtype: :class:`~certbot_nginx.obj.VirtualHost`
File:      ~/Code/certbot/certbot/certbot-nginx/certbot_nginx/configurator.py
Type:      instancemethod
```

Alternatively, you can do the same thing in the default Python interpreter (also after activating the venv):
```
$ python
Python 2.7.12 (default, Jul  1 2016, 15:12:24) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> from certbot_nginx.configurator import NginxConfigurator
>>> help(NginxConfigurator.choose_redirect_vhost)
```